lib/ansible/modules/pip.py

```diff
@@ -611,6 +613,61 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _normalize_vcs_packages(module: AnsibleModule,
+                            package_list: list[Package],
+                            pip: list[str]) -> list[Package]:
+    """Return list[Package] where all vcs URL packages are converted to actual package names"""
+    vcs_packages = [str(pkg) for pkg in package_list if _is_vcs_url(str(pkg))]
+
+    if not vcs_packages:
+        return package_list
+
+    other_packages = [pkg for pkg in package_list if not _is_vcs_url(str(pkg))]
+
+    # First, install as a dry-run report to get JSON output
+    frozen_env = os.environ.copy()
+    os.environ['TTY_COMPATIBLE'] = "0"
+    if os.environ.get('FORCE_COLOR'):
+        os.environ.pop('FORCE_COLOR')
+    rc, out, err = module.run_command([*pip, 'install', '--dry-run', '--ignore-installed', '--quiet', '--report=-', *vcs_packages])
+
+    if rc == 0:
+        # If it succeeds, handle the report
+        report = json.loads(out)
+        package_objects = [Package(install_report['metadata']['name'], version_string=install_report['metadata']['version'])
+                           for install_report in report['install']]
+    else:
+        # Else, if that fails due to --dry-run not being present in pip
+        # versions older than 22.2 (ex: Ubuntu 2204), fallback to using
+        # the potentially faulty pip download method and warn the user.
+
+        pip_downloads_dir = Path(module.tmpdir) / 'pipdownloads'
+        if pip_downloads_dir.exists():
+            shutil.rmtree(pip_downloads_dir)
+            pip_downloads_dir.mkdir()
+
+        rc, out, err = module.run_command([*pip, 'download', f'--dest={pip_downloads_dir}', '--no-deps', *vcs_packages])
+
+        if rc != 0:
+            # If it fails, just dump the error
+            os.environ.update(frozen_env)
+            module.fail_json(rc=rc, msg=out, err=err)
+
+        # Only from this point onward can we be sure that we're using the fallback, warn now.
+        module.warn("Using check_mode with vcs packages is potentially error prone on pip versions <22.2")
+
+        out_lines = out.splitlines()
+        saved_files = [Path(line.split(" ")[-1]).name for line in out_lines if 'Saved' in line]
+        package_names = [Path(name).stem for name in saved_files]
+
+        package_objects = [Package('-'.join(pkg.split('-')[:-1]),
+                                   version_string=pkg.split('-')[-1])
+                           for pkg in package_names]
```

Comment:
All these single-consumption lists are not necessary to keep in-memory. Let's use generation expressions:
```suggestion
        saved_files = (
            Path(line.split(" ")[-1]).name for line in out_lines
            if 'Saved' in line
        )
        package_names = (Path(name).stem for name in saved_files)

        package_objects = (
            Package(
                '-'.join(pkg.split('-')[:-1]),
                version_string=pkg.split('-')[-1],
            )
            for pkg in package_names
        )
```

Although, I see that the path objects are converted to strings and back to path. This seems excessive. Can't we just do

```suggestion
        saved_files = (
            Path(line.split(" ")[-1]) for line in out_lines
            if 'Saved' in line
        )
        package_stems = (name for name in saved_files)
        package_stem_pairs = (stem.split('-', 1) for stem in package_stems)

        package_objects = (
            Package(pkg[0], version_string=pkg[1])
            for pkg in package_stem_pairs
        )
```

Reply:
@pkingstonxyz why didn't you just accept the variant that avoids double path conversions?

Reply:
I ended up not accepting that variant because while it does avoid the double conversion, it doesn't work correctly. The bug is in stem.split('-', 1). This would split the package archive "pkg-with-long-name-0.0.1.zip" into ("pkg", "with-long-name-0.0.1.zip"). 

I did take the point about double conversions as going from `str` path -> `Path` -> `Path.name` -> `Path` -> `Path.stem` -> `stem.split('-')` is pointless and it can just as easily go `str` path -> `Path` -> `Path.stem` -> `stem.split('-')` as it is in line 662.

Reply:
Ah, interesting. It's best to document the train of thought by default right away. As a part of communicating your process and how you're handling the suggestions and so on.
If  not, some important bits might end up being missed by all the parties participating in the PR. Or this would trigger another round of back-and-forth comms that would take longer.

P.S. So it sounds like `stem.rsplit('-', -1)` would do what you intend, right?.

Reply:
Yes it looks like it, I had no idea there was an rsplit method on strings, yes that's perfect